src/main/flight/mixer.c

```diff
@@ -743,27 +744,118 @@ STATIC_UNIT_TESTED void servoMixer(void)
 
-        bool isFailsafeActive = failsafeIsActive();
+            if (rollPitchYawMix[i] > rollPitchYawMixMax) rollPitchYawMixMax = rollPitchYawMix[i];
+            if (rollPitchYawMix[i] < rollPitchYawMixMin) rollPitchYawMixMin = rollPitchYawMix[i];
+        }
+
+        // Scale roll/pitch/yaw uniformly to fit within throttle range
+        int16_t rollPitchYawMixRange = rollPitchYawMixMax - rollPitchYawMixMin;
+        int16_t throttleRange, throttle;
+        int16_t throttleMin = 0, throttleMax = 0;
+        static bool flightDirection3dReversed = false;
+
+        throttle = rcCommand[THROTTLE];
+
+        // Find min and max throttle based on condition
+        if (feature(FEATURE_3D)) {
+            static int16_t throttleMinPrevious = 0, throttleMaxPrevious = 0, throttlePrevious = 0;
+
+            if (rcData[THROTTLE] <= (rxConfig->midrc - flight3DConfig->deadband3d_throttle)) {
+                throttleMax = flight3DConfig->deadband3d_low;
+                throttleMin = escAndServoConfig->minthrottle;
+                flightDirection3dReversed = true;
+            } else if (rcData[THROTTLE] >= (rxConfig->midrc + flight3DConfig->deadband3d_throttle)) {
+                throttleMax = escAndServoConfig->maxthrottle;
+                throttleMin = flight3DConfig->deadband3d_high;
+                flightDirection3dReversed = false;
+            } else {
+                if (throttleMin) {
+                    throttleMax = throttleMaxPrevious;
+                    throttleMin = throttleMinPrevious;
+                    throttle = throttlePrevious;
+                } else {
+                    // when starting in neutral
+                    throttleMax = escAndServoConfig->maxthrottle;
+                    throttle = throttleMin = flight3DConfig->deadband3d_high;
+                }
+            }
+
+            throttleMaxPrevious = throttleMax;
+            throttleMinPrevious = throttleMin;
+            throttlePrevious = throttle;
+        } else {
+            throttleMin = escAndServoConfig->minthrottle;
+            throttleMax = escAndServoConfig->maxthrottle;
+        }
+
+        throttleRange = throttleMax - throttleMin;
+
+        if (rollPitchYawMixRange > throttleRange) {
+            motorLimitReached = true;
+            for (i = 0; i < motorCount; i++) {
+                rollPitchYawMix[i] = (rollPitchYawMix[i] * throttleRange) / rollPitchYawMixRange;
+            }
+        } else {
+            motorLimitReached = false;
+            throttleMin = throttleMin + (rollPitchYawMixRange / 2);
+            throttleMax = throttleMax - (rollPitchYawMixRange / 2);
+        }
+
+        // Now add in the desired throttle, but keep in a range that doesn't clip adjusted
+        // roll/pitch/yaw. This could move throttle down, but also up for those low throttle flips.
+        for (i = 0; i < motorCount; i++) {
+            motor[i] = rollPitchYawMix[i] + constrainf(throttle * currentMixer[i].throttle, throttleMin, throttleMax);
```

Comment:
`constrainf` is applied to integer variables

`motor[i]` is constrained again in following code, it may be better to mix bounds and apply constrain only once